### ISSUE TITLE
Distributed NCCL test suite doesn't work on Maxwells / test_all_gather_cuda (and possibly others) deadlock with CUDA 9 and system NCCL on latest nvidia/cuda images (driver 384.111)

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#5877]
Steps to reproduce:
1. Build a copy of PyTorch on nvidia/cuda Docker image
2. `mkdir /tmp/foo && mkdir /tmp/foo/barrier && mkdir /tmp/foo/test_dir`
3. Run `WORLD_SIZE=2 INIT_METHOD=env:// BACKEND=nccl TEMP_DIR=/tmp/foo python test/test_distributed.py TestDistBackend.test_all_gather_cuda`

Expected result: pass.

Actual result:

```
05:06:40 ======================================================================
05:06:40 FAIL: test_all_gather_cuda (__main__.TestDistBackend)
05:06:40 ----------------------------------------------------------------------
05:06:40 Traceback (most recent call last):
05:06:40   File "test_distributed.py", line 858, in wrapper
05:06:40     self._join_and_reduce(fn)
05:06:40   File "test_distributed.py", line 924, in _join_and_reduce
05:06:40     or first_process.exitcode == SKIP_IF_SMALL_WORLDSIZE_EXIT_CODE
05:06:40 AssertionError
05:06:40 
```

The assert is failing because process joining is timing out. I get a few left over Python processes and I have to `kill -9` them.

Sample log: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-cuda9-cudnn7-py2-test/4738//consoleFull (you can tell the processes deadlock because subsequent tests fail because they can't grab the address that is occupied by the deadlocked process.)

If you don't feel like building PyTorch, you can get a Docker image with a build with:

```
docker run -it -w /var/lib/jenkins --runtime=nvidia -e CUDA_VERSION=9 -e NVIDIA_VISIBLE_DEVICES=all ezyang/pytorch:distributed-fail /bin/bash
source /opt/conda/bin/activate
```

Don't forget to activate the conda environment. Then do instructions (2) and (3).

The problem does NOT occur on CUDA 8.

Also, running `apt-get remove -y libnccl-dev libnccl2` so that our bundled NCCL gets used also solves the problem.

CC @ngimel @teng-li

**Comment by teng-li** [comment]
I am taking a look at it.

**Comment by ngimel** [comment]
What GPUs is this on? On Maxwells some of the distributed tests are expected to fail (those that try to call nccl collectives for the tensors on the same GPU)

**Comment by ezyang** [comment]
These are g3.8xlarge's (https://aws.amazon.com/ec2/instance-types/) so it's Tesla M60.

**Comment by ezyang** [comment]
Oops, I'm a derp, Tesla is Maxwell architecture. So I guess what we need to do is detect if this case (using collectives for tensors on the same GPU) is happening and stop using nccl in this case.

**Comment by ngimel** [comment]
Something strange going on with multiprocessing and maxwells. When I comment out everything in _test_all_gather_multigpu_helper (and hardcode initialization to ```gloo```, though it does not matter), second rank still never reaches print after _build_tensor
```
WORLD_SIZE=2 INIT_METHOD=env:// BACKEND=nccl TEMP_DIR=/tmp/foo python test/test_distributed.py TestDistBackend.test_all_gather_multigpu
```
....
```
for dest in [0]:
        print("rank", rank)
        tensors = [_build_tensor(dest + 1).cuda(device=i)
                   for i in rankToGPUMapping[rank]]
        print("rank1", rank)
```

**Comment by ngimel** [comment]
I've deleted my previous message, my machine was in a strange state.

**Comment by ngimel** [comment]
The failing tests on Maxwells are test_broadcast_cuda, test_reduce_sum_cuda, test_all_reduce_sum_cuda, test_all_gather_cuda, all where the tensors are actually on the same GPU. Maxwell does not support pre-emption, so that's expected.

**Comment by ailzhang** [comment]
@ezyang We can change these tests to actually use multi gpus for testing. Currently we don't do it since we spawn 3 processes in Gloo test( it's 2 in NCCL tests), so we will need at least 3 GPUs to test it together with the new_group(which requires 3 processes) in Gloo. It's actually good to change it when we have >2 GPUs in our test environment. Is it already up online?

**Comment by ezyang** [comment]
Yep, we've got four GPU machines now. Right now `test_distributed.py` is being run in that configuration. https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-cuda8-cudnn6-py3-multigpu-test/

**Comment by ailzhang** [comment]
@ezyang This is great! I opened an issue to update the testcases.

**Comment by avmgithub** [comment]
Not sure if this is the same, on ppc64le I am seeing errors similar to ones described using Pascal (P100) ,  NVIDIA-SMI 384.81, cuda9

**Comment by avmgithub** [comment]
Additional info for ppc64le.  The tests works with Tesla V100-SXM2 with NVIDIA-SMI 390.31 and cuda 9.1

**Comment by ailzhang** [comment]
Closed by #6337
